### ISSUE TITLE
SerializationException Thrown When Attempting To Invoke Anonymous Method Referencing Static Members

### ISSUE THREAD

**Issue by martincostello** [dotnet/roslyn#1699]
*For this issue I have assumed that since Visual Studio 2015 was used that the Rosyln compiler was used by MSBuild to compile the code. If this is not the case, I'll re-log the bug against Visual Studio 2015 in Microsoft Connect.*

To replicate clone the ```https://github.com/martincostello/sqllocaldb.git``` repository as of commit ```803db560``` and run ```src\Build.cmd``` on a machine where only Visual Studio 2015 Ultimate Preview is installed (the specific version information copied from Visual Studio itself and Windows where this issue was found are at the bottom of this issue).

The ```Test``` target run after compilation will fail with 37 test failures caused by a ```System.Runtime.Serialization.SerializationException``` being thrown when attempting to call a method in a new AppDomain using an anonymous method which references static members.

An example stack trace from a failed test is below:

```
Test Name:	Extensions_GetConnectionStringForDefaultModel_Throws_If_Multiple_Connection_Strings_Are_Configured
Test FullName:	System.Data.SqlLocalDb.ExtensionsTests.Extensions_GetConnectionStringForDefaultModel_Throws_If_Multiple_Connection_Strings_Are_Configured
Test Source:	C:\Users\Martin\Source\Repos\sqllocaldb\src\SqlLocalDb.UnitTests\ExtensionsTests.cs : line 404
Test Outcome:	Failed
Test Duration:	0:00:00.0346031

Result Message:	Test method System.Data.SqlLocalDb.ExtensionsTests.Extensions_GetConnectionStringForDefaultModel_Throws_If_Multiple_Connection_Strings_Are_Configured threw exception System.Runtime.Serialization.SerializationException, but exception System.InvalidOperationException was expected. Exception message: System.Runtime.Serialization.SerializationException: Type 'System.Data.SqlLocalDb.ExtensionsTests+<>c__DisplayClass12' in assembly 'System.Data.SqlLocalDb.UnitTests, Version=1.15.0.0, Culture=neutral, PublicKeyToken=9a192a7522c9e1a0' is not marked as serializable.
Result StackTrace:	
at System.AppDomain.DoCallBack(CrossAppDomainDelegate callBackDelegate)
   at System.Data.SqlLocalDb.Helpers.InvokeInNewAppDomain(CrossAppDomainDelegate callBackDelegate, IDictionary`2 appDomainData, String configurationFile, String callerMemberName) in C:\Users\Martin\Source\Repos\sqllocaldb\src\SqlLocalDb.UnitTests\Helpers.cs:line 101
   at System.Data.SqlLocalDb.ExtensionsTests.Extensions_GetConnectionStringForDefaultModel_Throws_If_Multiple_Connection_Strings_Are_Configured() in C:\Users\Martin\Source\Repos\sqllocaldb\src\SqlLocalDb.UnitTests\ExtensionsTests.cs:line 406
```
Performing the same steps with Visual Studio 2013 and MSBuild 12.0 successfully complies the code and runs all the tests with no failures.

It appears that anonymous compiler-generated types associated with anonymous methods that only reference static members are no longer annotated with the ```[Serializable]``` attribute, causing the cross-AppDomain method invocations via ```Helpers.InvokeInNewAppDomain()``` to fail.

I would expect the same code compiled with Visual Studio 2015 to pass the same tests.

**Visual Studio 2015 Version Information**
```
Microsoft Visual Studio Ultimate 2015 Preview
Version 14.0.22310.1 DP
Microsoft .NET Framework
Version 4.6.00042

Installed Version: Ultimate

Architecture and Modeling Tools   00269-30000-00001-AA223
Microsoft Architecture and Modeling Tools
    
UML® and Unified Modeling Language™ are trademarks or registered trademarks of the Object Management Group, Inc. in the United States and other countries.

Visual Basic 2015 Preview   00269-30000-00001-AA223
Microsoft Visual Basic 2015 Preview

Visual C# 2015 Preview   00269-30000-00001-AA223
Microsoft Visual C# 2015 Preview

Visual C++ 2015 Preview   00269-30000-00001-AA223
Microsoft Visual C++ 2015 Preview

Visual F# 14 CTP   00269-30000-00001-AA223
Microsoft Visual F# 14 CTP

Visual Studio 2015 Preview Code Analysis Spell Checker   00269-30000-00001-AA223
Microsoft® Visual Studio® 2015 Preview Code Analysis Spell Checker

Portions of International CorrectSpell™ spelling correction system © 1993 by Lernout & Hauspie Speech Products N.V. All rights reserved.

The American Heritage® Dictionary of the English Language, Third Edition Copyright © 1992 Houghton Mifflin Company. Electronic version licensed from Lernout & Hauspie Speech Products N.V. All rights reserved.

Windows Phone SDK 8.0 - ENU   00269-30000-00001-AA223
Windows Phone SDK 8.0 - ENU

Application Insights Tools for Visual Studio Package   1.0
Application Insights Tools for Visual Studio

Common Azure Tools   1.3
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

ErrorList   1.0
Information about my package

Microsoft Azure Mobile Services Tools   1.3
Microsoft Azure Mobile Services Tools

Microsoft Code Digger   0.9
Microsoft Code Digger

Microsoft.Pex.VisualStudio   1.0
Pex

NuGet Package Manager   3.0.51103.210
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

PowerShell Tools   1.3
Provides file classification services using PowerShell

SQL Server Data Tools   14.0.41025.0
Microsoft SQL Server Data Tools
```

**Windows Version Information**
```
OS Name:	Microsoft Windows 10 Pro Technical Preview
Version:	10.0.10049 Build 10049
```

**Comment by martincostello** [comment]
Having since updated Visual Studio 2015 Preview to CTP6, the tests now all pass as expected.

**Comment by theoy** [comment]
Yep! We caught this internally as well. It turns out that while we never specced the serializability of lambda delegate instances... lambdas that captured empty closures were implicitly serializable in the past... so we had to ensure that again even though we are not generating them as instance methods.

However! As an added perk, in the JITs that we ran on, we observed that @vsadov's new lambda generation shaves off about ~30-35% of the delegate invocation overhead. Woo! :)
